Add Functions Support, and new Models from Jun 13th Update

hansjm10 opened the pull request:
This commit adds in function support as described in the latest June 13th update. It also adds the new models which support functions. This closes #146

tkoenig89 commented:
This looks pretty promising! Thank you for sharing. 

I'm just trying it out and found something that looks like a typo to me:

![image](https://github.com/OkGoDoIt/OpenAI-API-dotnet/assets/1177652/3a754d7a-f189-4e0d-8f29-f9af55a428dd)

As I only add a single function, i would expect the class to be named "Function". Or do I maybe miss a point of the class? 

Either way, thanks again for adding this. I hope this gets merged here as Microsoft is not supporting functions calls in their new OpenAI package as of right now.

ClusterM commented:
What about chat endpoint requests? And function parameters description as `JObject` is not good idea.

hansjm10 commented:
> What about chat endpoint requests? 
I figured chat endpoint requests can come later as this is largely adding in the backend for function support. 
> function parameters description as `JObject` is not good idea.
Do you have any suggestions as I don't like it very much either. The model is expecting some sort of JSON object and fails if you send JSON serialized as a string.

tkoenig89 commented:
I was thinking about a way to pass a type instead of the parameters and the
Function would take the type information from that type and generate
Jsonschema from it. I saw already that Newtonsoft has Classes to generate
JsonSchema but did not had the time to look into it. The nice thing about
passing a type would be, that you can later use that type to deserialize
the function call from ChatGpt. This is not realy thought through at the
moment. Just some brainstorming i made a few hours ago.


Jordan Hans ***@***.***> schrieb am Fr., 16. Juni 2023, 19:04:

> What about chat endpoint requests?
> I figured chat endpoint requests can come later as this is largely adding
> in the backend for function support.
> function parameters description as JObject is not good idea.
> Do you have any suggestions as I don't like it very much either. The model
> is expecting some sort of JSON object and fails if you send JSON serialized
> as a string.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OkGoDoIt/OpenAI-API-dotnet/pull/149#issuecomment-1594991242>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI7QNHM4TKGPM5QEPBQETTXLSG3JANCNFSM6AAAAAAZIVXZ4I>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

tkoenig89 commented:
Could not do more testing atm. But this seems to work fine, at least till serialization:

Function.cs
```C#
[JsonProperty("parameters", Required = Required.Default)]
public JSchema Parameters { get; set; }

public Function(string name, string description, Type type)
{
    this.Name = name;
    this.Description = description;
    this.Parameters = new JSchemaGenerator().Generate(type); // <--
}
```

Test
```C#

  private class TestClass
  {
    public string TestString { get; set; }
    public int TestInt { get; set; }
    public bool TestBool { get; set; }
  }
  
  [Test]
  public void TestFunctionWithSchema(){
    var fn = new Function("s","d",typeof(TestClass));
    var str = JsonConvert.SerializeObject(fn);
  }
```

hansjm10 commented:
Unfortunately It looks like Newtonsofts JsonSchema licensing is AGPL and only supports up to 1000 requests per hour without purchasing it. However it is worth exploring the idea of easily passing in the parameters schema using a Type. I previously was using NJsonSchema for this but didn't like the idea of including outside packages, that weren't already being used in the project. 

The benefit to using the JObject is we can easily convert other formats to it. The downside is it isn't an easy way to natively generate the schema. I thought about including a basic schema generator but felt it was outside the scope of the PR and project as a whole.

tkoenig89 commented:
I agree on keeping external packages to a minimum. With some AI support i came up with a simple JSON Schema generator, that might help here. I just did a first test with it, but as its getting pretty late here i cannot realy concentrate very good :D

Moved code to a gist, so i dont blow up the discussion here.
https://gist.github.com/tkoenig89/e35d6ffc2979746476893fe00234ab30

 If this looks like a way to go. I'm glad to provide a cleaned up version with a bunch of tests. After some sleep :)

ClusterM commented:
There is my solution:
```C#
    public interface IJsonSchema
    {
        string Type { get; }
    }

    public class JsonObjectSchema : IJsonSchema
    {
        public string Type { get; } = "object";
        public required Dictionary<string, IJsonSchema> Properties { get; set; }
        public List<string>? Required { get; set; }
    }

    public class JsonArraySchema : IJsonSchema
    {
        public string Type { get; } = "array";
        public required IJsonSchema Items { get; set; }
        public int? MinItems { get; set; }
        public int? MaxItems { get; set; }
        public bool? UniqueItems { get; set; }
    }

    public class JsonStringSchema : IJsonSchema
    {
        public string Type { get; } = "string";
        public string? Description { get; set; }
        public List<string>? Enum { get; set; }
        public string? Pattern { get; set; }
        public int? MinLength { get; set; }
        public int? MaxLength { get; set; }
    }

    public class JsonNumberSchema : IJsonSchema
    {
        public string Type { get; } = "number";
        public string? Description { get; set; }
        public double? Minimum { get; set; }
        public double? Maximum { get; set; }
        public double? MultipleOf { get; set; }
        public double? ExclusiveMaximum { get; set; }
        public double? ExclusiveMinimum { get; set; }
    }

    public class JsonBooleanSchema : IJsonSchema
    {
        public string Type { get; } = "boolean";
        public string? Description { get; set; }
    }
```

Usage:
```C#
var mySchema = new JsonObjectSchema
{
    Properties = new Dictionary<string, IJsonSchema>
    {
        ["name"] = new JsonStringSchema
        {
            Description = "The name of the person",
            Enum = new List<string> {"Alice", "Bob", "Charlie"},
            MinLength = 1,
            MaxLength = 100
        },
        ["age"] = new JsonNumberSchema
        {
            Description = "The age of the person",
            Minimum = 0,
            Maximum = 150,
            MultipleOf = 1,
            ExclusiveMaximum = false,
            ExclusiveMinimum = false
        },
        ["hobbies"] = new JsonArraySchema
        {
            Items = new JsonStringSchema(),
            MinItems = 0,
            MaxItems = 100,
            UniqueItems = true
        },
        ["is_married"] = new JsonBooleanSchema
        {
            Description = "Marital status of the person"
        }
    },
    Required = new List<string> {"name", "age"}
};

var options = new JsonSerializerOptions { WriteIndented = true, DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull };
var jsonString = JsonSerializer.Serialize(mySchema, options);

Console.WriteLine(jsonString);
```

It can serialize, but not deserialize. But deserialization is not required.

ClusterM commented:
There is my solution:
```C#
public interface IJsonSchema
{
    string Type { get; }
}

public class JsonObjectSchema : IJsonSchema
{
    [JsonPropertyName("type")]
    public string Type { get; } = "object";

    [JsonPropertyName("properties")]
    public required Dictionary<string, IJsonSchema> Properties { get; set; }

    [JsonPropertyName("required")]
    public List<string>? Required { get; init; }
}

public class JsonArraySchema : IJsonSchema
{
    [JsonPropertyName("type")]
    public string Type { get; } = "array";

    [JsonPropertyName("items")]
    public required IJsonSchema Items { get; set; }

    [JsonPropertyName("minItems")]
    public int? MinItems { get; init; }

    [JsonPropertyName("maxItems")]
    public int? MaxItems { get; init; }

    [JsonPropertyName("uniqueItems")]
    public bool? UniqueItems { get; init; }

    [JsonPropertyName("description")]
    public string? Description { get; init; }
}

public class JsonStringSchema : IJsonSchema
{
    [JsonPropertyName("type")]
    public string Type { get; } = "string";

    [JsonPropertyName("description")]
    public string? Description { get; init; }

    [JsonPropertyName("enum")]
    public List<string>? Enum { get; init; }

    [JsonPropertyName("pattern")]
    public string? Pattern { get; init; }

    [JsonPropertyName("minLength")]
    public int? MinLength { get; init; }

    [JsonPropertyName("maxLength")]
    public int? MaxLength { get; init; }
}

public class JsonNumberSchema : IJsonSchema
{
    [JsonPropertyName("type")]
    public string Type { get; } = "number";

    [JsonPropertyName("description")]
    public string? Description { get; init; }

    [JsonPropertyName("minimum")]
    public double? Minimum { get; init; }

    [JsonPropertyName("maximum")]
    public double? Maximum { get; init; }

    [JsonPropertyName("multipleOf")]
    public double? MultipleOf { get; init; }

    [JsonPropertyName("exclusiveMaximum")]
    public double? ExclusiveMaximum { get; init; }

    [JsonPropertyName("exclusiveMinimum")]
    public double? ExclusiveMinimum { get; init; }
}

public class JsonBooleanSchema : IJsonSchema
{
    [JsonPropertyName("type")]
    public string Type { get; } = "boolean";

    [JsonPropertyName("description")]
    public string? Description { get; init; }
}

public class JsonSchemaConverter : JsonConverter<IJsonSchema>
{
    public override IJsonSchema? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        throw new NotImplementedException();
    }

    public override void Write(Utf8JsonWriter writer, IJsonSchema value, JsonSerializerOptions options)
    {
        JsonSerializer.Serialize(writer, value, value.GetType(), options);
    }
}
```

Usage:
```C#
var mySchema = new JsonObjectSchema
{
    Properties = new Dictionary<string, IJsonSchema>
    {
        ["name"] = new JsonStringSchema
        {
            Description = "The name of the person",
            Enum = new List<string> { "Alice", "Bob", "Charlie" },
            MinLength = 1,
            MaxLength = 100
        },
        ["age"] = new JsonNumberSchema
        {
            Description = "The age of the person",
            Minimum = 0,
            Maximum = 150,
            MultipleOf = 1
        },
        ["hobbies"] = new JsonArraySchema
        {
            Items = new JsonStringSchema()
            {
                MaxLength = 30
            },
            MinItems = 0,
            MaxItems = 100,
            UniqueItems = true
        },
        ["is_married"] = new JsonBooleanSchema
        {
            Description = "Marital status of the person"
        }
    },
    Required = new List<string> { "name", "age" }
};

var options = new JsonSerializerOptions { WriteIndented = true, DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull };
options.Converters.Add(new JsonSchemaConverter());
var jsonString = JsonSerializer.Serialize(mySchema, options);

Console.WriteLine(jsonString);
```

Output:
```JSON
{
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "The name of the person",
      "enum": [
        "Alice",
        "Bob",
        "Charlie"
      ],
      "minLength": 1,
      "maxLength": 100
    },
    "age": {
      "type": "number",
      "description": "The age of the person",
      "minimum": 0,
      "maximum": 150,
      "multipleOf": 1
    },
    "hobbies": {
      "type": "array",
      "items": {
        "type": "string",
        "maxLength": 30
      },
      "minItems": 0,
      "maxItems": 100,
      "uniqueItems": true
    },
    "is_married": {
      "type": "boolean",
      "description": "Marital status of the person"
    }
  },
  "required": [
    "name",
    "age"
  ]
}
```

It can serialize, but not deserialize. But deserialization is not required.

hansjm10 commented:
I have done some adjustments to the parameters type and setter

```
 [JsonProperty("parameters", Required = Required.Default)]
        public object Parameters
        {
            get
            {
                return _parameters;
            }
            set
            {
                try
                {
                    if (value is string jsonStringValue)
                    {
                        _parameters = JObject.Parse(jsonStringValue);
                    }
                    if (value is JObject jObjectValue)
                    {
                        _parameters = jObjectValue;
                    }
                    else
                    {
                        var settings = new JsonSerializerSettings
                        {
                            NullValueHandling = NullValueHandling.Ignore
                        };
                        var jsonString = JsonConvert.SerializeObject(value, settings);
                        _parameters = JObject.Parse(jsonString);
                    }
                }
                catch (JsonException e)
                {
                    throw new ArgumentException("Could not convert the provided object into a JSON object. Make sure that the object is serializable and its structure matches the required schema.", e);
                }
            }
        }
```

This avoid implementing our own JsonSchema as the user will be able to provide their own which is arguably going to be better than anything we need to implement and support in the future. If the user wants to use a schema like Newtonsoft, NJsonSchema, or their own implementation it can handle it, if the user passes in a dictionary<string, obj> it can handle it, or if the user passes in a string formatted as JSON it can handle it. 

@ClusterM I tested this using your implementation (adjusting it slightly for support with Newtonsoft over System.Text.Json) and the original JsonObjectSchema and Parameters were identical.

If the user doesn't provide the needed schema they will receive a 400 error from the OpenAI endpoint.

Do we have any objections to this handling?

tkoenig89 commented:
Sounds reasonable to me. Using object instead of JObject makes it more
accesible and less dependent on Newtonsoft.

Jordan Hans ***@***.***> schrieb am Sa., 17. Juni 2023, 18:06:

> I have done some adjustments to the parameters type and setter
>
>  [JsonProperty("parameters", Required = Required.Default)]
>         public object Parameters
>         {
>             get
>             {
>                 return _parameters;
>             }
>             set
>             {
>                 try
>                 {
>                     if (value is string jsonStringValue)
>                     {
>                         _parameters = JObject.Parse(jsonStringValue);
>                     }
>                     if (value is JObject jObjectValue)
>                     {
>                         _parameters = jObjectValue;
>                     }
>                     else
>                     {
>                         var settings = new JsonSerializerSettings
>                         {
>                             NullValueHandling = NullValueHandling.Ignore
>                         };
>                         var jsonString = JsonConvert.SerializeObject(value, settings);
>                         _parameters = JObject.Parse(jsonString);
>                     }
>                 }
>                 catch (JsonException e)
>                 {
>                     throw new ArgumentException("Could not convert the provided object into a JSON object. Make sure that the object is serializable and its structure matches the required schema.", e);
>                 }
>             }
>         }
>
> This avoid implementing our own JsonSchema as the user will be able to
> provide their own which is arguably going to be better than anything we
> need to implement and support in the future. If the user wants to use a
> schema like Newtonsoft, NJsonSchema, or their own implementation it can
> handle it, if the user passes in a dictionary<string, obj> it can handle
> it, or if the user passes in a string formatted as JSON it can handle it.
>
> @ClusterM <https://github.com/ClusterM> I tested this using your
> implementation (adjusting it slightly for support with Newtonsoft over
> System.Text.Json) and the original JsonObjectSchema and Parameters were
> identical.
>
> If the user doesn't provide the needed schema they will receive a 400
> error from the OpenAI endpoint.
>
> Do we have any objections to this handling?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OkGoDoIt/OpenAI-API-dotnet/pull/149#issuecomment-1595795305>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAI7QNDLH2BK7TAGXJHMIRDXLXIY3ANCNFSM6AAAAAAZIVXZ4I>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

tkoenig89 commented:
@hansjm10 you might want to use `JObject.FromObject(value)` instead of serializing to string and parsing as JObject, At least this seemed to be the simplest solution i used when i tested the original version of the PR.

ErikDombi commented:
I can't get this PR to work when streaming the chat completion. Anyone else having issues with this, or am I just doing something wrong?

hansjm10 commented:
> @ErikDombi  I can't get this PR to work when streaming the chat completion. Anyone else having issues with this, or am I just doing something wrong?

Mind sending what you have tried and your response from OpenAI? I am able to use 

~~~
   IAsyncEnumerable<string> asyncEnumerable = conversation.StreamResponseEnumerableFromChatbotAsync();

   await foreach(var res in asyncEnumerable)
   {
	Console.WriteLine(res);
   }
~~~

and the returned streamed results are correct

brondavies commented on OpenAI_Tests/ChatEndpointTests.cs:
```diff
@@ -149,6 +153,54 @@ public void ChatBackAndForth(string model)
+				var response = await conversation.GetResponseFromChatbotAsync();
+				
+
+				Assert.AreEqual(response, "The current weather in Boston is sunny with a temperature of 22 degrees Celsius.");
```
These parameters are reversed. The expected response should come first.
```suggestion
				Assert.AreEqual("The current weather in Boston is sunny with a temperature of 22 degrees Celsius.", response);
```

brondavies reviewed (changes requested):
In order for this PR to work as the example code does, You'll have to include this change in [ChatMessage.cs](https://github.com/OkGoDoIt/OpenAI-API-dotnet/pull/149/files#diff-529524c4809e4dba9e3a7cbe2aa9fdf2d68e8a68402d79e796c154c76d6f6597)

```csharp
		[JsonProperty("content", NullValueHandling = NullValueHandling.Include)]
		public string Content { get; set; }
```
This is because the function response message from OpenAI has a result with a content of `null`.  If it is not included in the message list on subsequent requests, the API will give you an error:  

```json
"error": {
    "message": "'content' is a required property - 'messages.1'",
    "type": "invalid_request_error",
    "param": null,
    "code": null
  }
```

brondavies commented:
Here is an updated test case that follows the workflow in the example code from OpenAI. See the complete python example at https://platform.openai.com/docs/guides/gpt/function-calling

```csharp

		public async Task SummarizeFunctionResult()
		{
			try
			{
				var api = new OpenAI_API.OpenAIAPI();
                var functionList = new List<Function>
                {
                    BuildFunctionForTest()
                };
				var conversation = api.Chat.CreateConversation(new ChatRequest { 
					Model = Model.ChatGPTTurbo0613,
					Functions = functionList
				});
				conversation.AppendUserInput("What is the weather like in Boston?");

                var response = await conversation.GetResponseFromChatbotAsync();

				Assert.IsNull(response);

                var functionMessage = new ChatMessage
                {
                    Role = ChatMessageRole.Function,
                    Name = "get_current_weather",
                    Content = "{\"temperature\": \"22\", \"unit\": \"celsius\", \"description\": \"sunny\"}"
                };
                conversation.AppendMessage(functionMessage);
				response = await conversation.GetResponseFromChatbotAsync();
				
				Assert.AreEqual("The current weather in Boston is sunny with a temperature of 22 degrees Celsius.", response);
			}
			catch(NullReferenceException ex)
			{
				Console.WriteLine(ex.Message, ex.StackTrace);
				Assert.False(true);
			}
        }
```

brondavies commented on OpenAI_Tests/ChatEndpointTests.cs:
```diff
@@ -149,6 +153,54 @@ public void ChatBackAndForth(string model)
+		{
+			try
+			{
+				var api = new OpenAI_API.OpenAIAPI();
+
+				var conversation = api.Chat.CreateConversation(new ChatRequest { Model = Model.ChatGPTTurbo0613 });
+				conversation.AppendUserInput("What is the weather like in Boston?");
+
+                var initialAssistantMessage = new ChatMessage
+                {
+                    Role = ChatMessageRole.Assistant,
+                    Content = "(null)",
+                    Function_Call = new Function_Call()
+                    {
+                        Name = "get_current_weather",
+                        Arguments = "{ \"location\": \"Boston, MA\"}"
+                    }
+                };
+                conversation.AppendMessage(initialAssistantMessage);
+                var functionMessage = new ChatMessage
+                {
+                    Role = ChatMessageRole.Function,
+                    Name = "get_current_weather",
+                    Content = "{\"temperature\": \"22\", \"unit\": \"celsius\", \"description\": \"Sunny\"}"
+                };
+                var functionList = new List<Function>
+                {
+                    BuildFunctionForTest()
+                };
+				conversation.AppendMessage(functionMessage);
+                conversation.RequestParameters.Functions = functionList;
+				conversation.RequestParameters.Temperature = 0;
+
+
+				var response = await conversation.GetResponseFromChatbotAsync();
+				
```
```suggestion
                var api = new OpenAI_API.OpenAIAPI();
                var functionList = new List<Function>
                {
                    BuildFunctionForTest()
                };
                var conversation = api.Chat.CreateConversation(new ChatRequest { 
                    Model = Model.ChatGPTTurbo0613,
                    Functions = functionList
                });
                conversation.AppendUserInput("What is the weather like in Boston?");

                var response = await conversation.GetResponseFromChatbotAsync();

                Assert.IsNull(response);

                var functionMessage = new ChatMessage
                {
                    Role = ChatMessageRole.Function,
                    Name = "get_current_weather",
                    Content = "{\"temperature\": \"22\", \"unit\": \"celsius\", \"description\": \"sunny\"}"
                };
                conversation.AppendMessage(functionMessage);
                response = await conversation.GetResponseFromChatbotAsync();
```

62316e commented:
Is there any example how to extract function name and arguments from GetResponseFromChatbotAsync?

`conversation.AppendUserInput("What is the weather like in Boston?");` returns null.

hansjm10 commented:
> Is there any example how to extract function name and arguments from GetResponseFromChatbotAsync?
> 
> `conversation.AppendUserInput("What is the weather like in Boston?");` returns null.

The newest commit adds in new chat endpoints. You can use 
~~~
var response = GetFunction_CallResponseAsync()
~~~

and access function names and arguments like

~~~
var name = response.Name;

var arguments = response.Arguments;
~~~

hansjm10 commented:
> CallResponseAsync should return either message or function, since you don't always know if you need to call a function or not

Going to take some time later to think about this situation and see how the official documentation handles this. FWIW You can easily check if the response is null for a function_call if your expecting it.

~~~
var response = await conversation.GetFunction_CallResponseAsync();
// The user input didn't trigger a function call.
if(response == null)
{
    Console.WriteLine(conversation.Messages.Last().content)
}
~~~

as with streaming a function that is a bit weird since it supposed to be for internal use for accessing another API. If you could come up with a use case I can look into it more. That being said I'm not to familiar with the streaming side if somebody else wants to look into it too and see if there is a good solution.

62316e commented:
> > CallResponseAsync should return either message or function, since you don't always know if you need to call a function or not
> 
> Going to take some time later to think about this situation and see how the official documentation handles this. FWIW You can easily check if the response is null for a function_call if your expecting it.
> 
> ```
> var response = await conversation.GetFunction_CallResponseAsync();
> // The user input didn't trigger a function call.
> if(response == null)
> {
>     Console.WriteLine(conversation.Messages.Last().content)
> }
> ```
> 
> as with streaming a function that is a bit weird since it supposed to be for internal use for accessing another API. If you could come up with a use case I can look into it more. That being said I'm not to familiar with the streaming side if somebody else wants to look into it too and see if there is a good solution.

I did some hacks on my side but would like to see the production ready PR. Streaming is also working on my side (with hacks). 

Use case (chatbot) with streaming:
- User: "tell me a joke" <- No function call is required here and system can stream the response directly to the user
- User: "what's the weather" <- System checks for function call and makes a second call (also with streaming) and stream is returned to the user.

In both cases there should be a single CallResponseAsync method which returns function or message. Because you don't know in advance if function will be called or not.

hansjm10 commented:
> In both cases there should be a single CallResponseAsync method which returns function or message. Because you don't know in advance if function will be called or not.

These are two different return types. If you are expecting the possibility of a function you should be checking if the resulted response is a function anyway. If you want more control over this you can use `CreateChatCompletionAsync` which returns the ChatResult.

brondavies commented on OpenAI_API/Chat/Conversation.cs:
```diff
@@ -111,34 +112,51 @@ public void AppendMessage(ChatMessage message)
+        /// Calls the API to get a response and retrieves the <see cref="Function_Call"/> from the resulting <see cref="ChatMessage"/>.
+        /// </summary>
+        /// <returns>The Function_Call object from the chat message.</returns>
+        public async Task<Function_Call> GetFunction_CallResponseAsync()
+        {
+            return await GetFromChatResultAsync(message => message.Function_Call);
+        }
```
This seems like the wrong way to implement this because it implies you know that the API is going to request a function call based on the input.  The way it's implemented in the python library is to check for an empty `content` return from `GetResponseFromChatbotAsync` and then get `response["choices"][0]["function_call"]` from the last response to check the named Function_Call.

hansjm10 replied:
Not sure why but I'm not able to find any reference to functions in the openai/openai-python library, if you could link what your going off of I would appreciate it. The cookbook utilizes JSON for examples which we shouldn't be working with as our end result. 

If function_call is not null that implies the existence of a function,. The real question is should we have the user handle the function call or the wrapper handle it. 

If it is null we can just output the response as a string like you said earlier. This is just rambling but we could have the user pass in a list of Func's, then invoke them based on the response. (Which in turn would finally call the summarize). This would circumvent needing to return function_call to the user directly. 

That seems a bit advanced and could be outside the scope but it's an interesting thought.

Alternatively this is all handled just by using 'CreateChatCompletionAsync' and letting the user manually check function_call + content.

brondavies replied:
I would go for the simplest implementation at this point.  I see your point that there really isn't a reference implementation in the python library. It's just relying on the dictionary.  Since there are a dozen ways someone might want to implement handling the presence of function_call in the return - I think the way it works now will suffice.

hansjm10 commented on OpenAI_Tests/ChatEndpointTests.cs:
```diff
@@ -149,6 +153,44 @@ public void ChatBackAndForth(string model)
+                };
+                var conversation = api.Chat.CreateConversation(new ChatRequest { 
+                    Model = Model.ChatGPTTurbo0613,
+                    Functions = functionList
```
You don't like the test succeeding one attempt and failing another with 0 changes?

brondavies commented:
@hansjm10 Can you also add the following method to [Conversation.cs](https://github.com/hansjm10/OpenAI-API-dotnet/blob/master/OpenAI_API/Chat/Conversation.cs) ?
I think this would make the solution more complete.
```csharp
        /// <summary>
        /// Creates and appends a <see cref="ChatMessage"/> to the chat history with the Role of <see cref="ChatMessageRole.Function"/>.  The function message is a response to a request from the system for output from a predefined function.
        /// </summary>
        /// <param name="functionName">The name of the function for which the content has been generated as the result</param>
        /// <param name="content">The text content (usually JSON)</param>
        public void AppendFunctionMessage(string functionName, string content) => AppendMessage(new ChatMessage(ChatMessageRole.Function, content) { Name = functionName });

```

cnSchwarzer commented:
I just can't believe this function still stuck at PR after two weeks.
Also, `Function_Call` violates almost all naming rules in C#.

logikonline commented:
Some feedback on usability of the functions:

Not a fan of "GetResponseFromChatbotAsync" returning null when the action has choices.  Would like to see another routine added called "GetChoicesFromChatbotAsync" with the array of Choices since the likelihood of using functions will only increase over time & using the first message of Choice[0] feels hinky requiring the need to go back to the Conversation object to access "MostRecentApiResult".

`
		/// <summary>
		/// Calls the API to get a response, which is appended to the current chat's <see cref="Choices"/> as an <see cref="ChatMessageRole.Assistant"/> <see cref="ChatMessage"/>.
		/// </summary>
		/// <returns>The Choices of the response from the chatbot API</returns>
		public async Task<IReadOnlyList<ChatChoice>> GetChoicesFromChatbotAsync()
		{
			ChatRequest req = new ChatRequest(RequestParameters);
			req.Messages = _Messages.ToList();

			var res = await _endpoint.CreateChatCompletionAsync(req);
			MostRecentApiResult = res;

			if (res != null && res.Choices != null)
				return res.Choices;
			return null;
		}
`

Overall, I hope this gets merged sooner than later since it is functional

brondavies commented:
> Not a fan of "GetResponseFromChatbotAsync" returning null when the action has choices.

@logikonline I hear what you're saying although there is already a method `CreateChatCompletionAsync` which gives you full access to the response.  `GetResponseFromChatbotAsync` is more of a convenience function for strictly implementing basic chat functionality.  A more advanced use case requiring access to functions makes that just another approach for the same functionality.

brondavies commented on OpenAI_API/Chat/ChatMessage.cs:
```diff
@@ -48,16 +49,22 @@ public ChatMessageRole Role
+        /// The name and arguments of a function that should be called, as generated by the model.
+        /// </summary>
+        [JsonProperty("function_call")]
+        public Function_Call Function_Call { get; set; }
```
We can follow standard naming convention here and still use the json name as-is - more changes ensue down the line...
```suggestion
        public FunctionCall FunctionCall { get; set; }
```

brondavies commented on OpenAI_API/Chat/ChatRequest.cs:
```diff
@@ -118,11 +119,27 @@ public string StopSequence
+        /// </summary>
+        /// <remarks>
+        /// If this field is not specified, the default behavior ("auto") allows the model to automatically decide whether to call a function or not.
+        /// Specify the name of the function to call in the "Name" attribute of the Function_Call object.
+        /// If you do not want the model to call any function, pass "None" for the "Name" attribute.
+        /// </remarks>
+        [JsonProperty("function_call")]
+        [JsonConverter(typeof(FunctionCallConverter))]
+        public Function_Call Function_Call { get; set; }
```
```suggestion
        /// Specify the name of the function to call in the "Name" attribute of the FunctionCall object.
        /// If you do not want the model to call any function, pass "None" for the "Name" attribute.
        /// </remarks>
        [JsonProperty("function_call")]
        [JsonConverter(typeof(FunctionCallConverter))]
        public FunctionCall FunctionCall { get; set; }
```

brondavies commented on OpenAI_API/Chat/ChatRequest.cs:
```diff
@@ -144,6 +161,8 @@ public ChatRequest(ChatRequest basedOn)
 			this.PresencePenalty = basedOn.PresencePenalty;
 			this.LogitBias = basedOn.LogitBias;
+			this.Functions = basedOn.Functions;
+			this.Function_Call = basedOn.Function_Call;
```
```suggestion
			this.FunctionCall = basedOn.FunctionCall;
```

brondavies commented on OpenAI_API/ChatFunctions/FunctionCallConverter.cs:
```diff
@@ -0,0 +1,46 @@
+    {
+        public override bool CanConvert(Type objectType)
+        {
+            return (objectType == typeof(Function_Call));
+        }
+
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            var functionCall = value as Function_Call;
+
+            if (functionCall.Name == "none" || functionCall.Name == "auto")
+            {
+                serializer.Serialize(writer, functionCall.Name);
+            }
+            else
+            {
+                serializer.Serialize(writer, functionCall);
+            }
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader.TokenType == JsonToken.String)
+            {
+                var functionCallType = (string)serializer.Deserialize(reader, typeof(string));
+
+                if (functionCallType == "none" || functionCallType == "auto")
+                {
+                    return new Function_Call { Name = functionCallType };
+                }
+            }
+            else if (reader.TokenType == JsonToken.StartObject)
+            {
+                return serializer.Deserialize<Function_Call>(reader);
+            }
+
+            throw new ArgumentException("Unsupported type for Function_Call");
```
```suggestion
            return (objectType == typeof(FunctionCall));
        }

        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
        {
            var functionCall = value as FunctionCall;

            if (functionCall.Name == "none" || functionCall.Name == "auto")
            {
                serializer.Serialize(writer, functionCall.Name);
            }
            else
            {
                serializer.Serialize(writer, functionCall);
            }
        }

        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
        {
            if (reader.TokenType == JsonToken.String)
            {
                var functionCallType = (string)serializer.Deserialize(reader, typeof(string));

                if (functionCallType == "none" || functionCallType == "auto")
                {
                    return new FunctionCall { Name = functionCallType };
                }
            }
            else if (reader.TokenType == JsonToken.StartObject)
            {
                return serializer.Deserialize<FunctionCall>(reader);
            }

            throw new ArgumentException("Unsupported type for FunctionCall");
```

brondavies commented on OpenAI_API/ChatFunctions/Function_Call.cs:
```diff
@@ -0,0 +1,24 @@
+    /// <summary>
+    /// An optional class to be used with models that support returning function calls.
+    /// </summary>
+    public class Function_Call
```
You would also want to rename this file to FunctionCall.cs
```suggestion
    public class FunctionCall
```

brondavies reviewed (changes requested):
Just renaming `Function_Call` to `FunctionCall` for compliance with recommended C# naming conventions.

logikonline commented:
@brondavies I am not hung up so much on the name, only the functionality - the change is fine to me.

I prefer the "GetResultFromChatbotAsync" only because it stays inline with the workflow of the Conversation object.

I would maybe make a convenience enum to the ChatResult called "NextAction" which makes the flow even more developer friendly.

````
	public enum Action { Stop, EOS, Function, Unknown }


        public class ChatResult : ApiResultBase
        {
              ...
              public Action NextAction { get; set; }
              ...
        }
        
        public async Task<ChatResult> GetResultFromChatbotAsync()
        {
	        ChatRequest req = new ChatRequest(RequestParameters);
	        req.Messages = _Messages.ToList();
        
	        var res = await _endpoint.CreateChatCompletionAsync(req);
	        MostRecentApiResult = res;
	        if (res != null)
	        {
		        if (res.Choices.Count > 0)
		        {
			        switch (res.Choices[0].FinishReason.ToLower())
			        {
				        case "stop":
					        res.NextAction = Action.Stop;
					        break;
				        case "eos":
					        res.NextAction = Action.EOS;
					        break;
				        case "function_call":
					        res.NextAction = Action.Function;
					        break;
				        default:
					        res.NextAction = Action.Unknown;
					        break;
			        }
		        }
		        else
		        {
			        res.NextAction = Action.Unknown;
		        }
	        }
	        return res;
        }

````

AleksandrFurmenkovOfficial commented:
Can someone clarify what is the proper way to work with StreamResponseEnumerableFromChatbotAsync()?

brondavies commented:
@OkGoDoIt I think this is ready to merge

varon commented on OpenAI_API/ChatFunctions/Function.cs:
```diff
@@ -0,0 +1,101 @@
+    public class Function
+    {
+        /// <summary>
+        /// The name of the function to be called. Must be a-z, A-Z, 0-9, or contain underscores and dashes, with a maximum length of 64.
```
Are we providing any validation on this? Can we match it against a regex before sending off to OpenAI?

varon commented on OpenAI_API/ChatFunctions/Function.cs:
```diff
@@ -0,0 +1,101 @@
+            }
+            set
+            {
+                try
+                {
+                    if (value is string jsonStringValue)
+                    {
+                        _parameters = JObject.Parse(jsonStringValue);
+                    }
+                    else if (value is JObject jObjectValue)
+                    {
+                        _parameters = jObjectValue;
+                    }
+                    else
+                    {
+                        var settings = new JsonSerializerSettings
+                        {
+                            NullValueHandling = NullValueHandling.Ignore
+                        };
+                        _parameters = JObject.FromObject(value, JsonSerializer.Create(settings));
+                    }
+                }
+                catch (JsonException e)
+                {
+                    throw new ArgumentException("Could not convert the provided object into a JSON object. Make sure that the object is serializable and its structure matches the required schema.", e);
+                }
```
Might be worth placing this in a helper method for clarity/simplicity.

varon reviewed (approved):
This looks good. Some minor nitpicks, but those can be fixed up later. Will be great to get this in.

varon commented:
@OkGoDoIt ^

ecsplendid commented:
Got this error running in Unity:

```
Sender:System.Threading.Tasks.Task`1[System.Threading.Tasks.VoidTaskResult] Exception:System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. (Error creating 'OpenAI_API.ChatFunctions.FunctionCallConverter'.) ---> Newtonsoft.Json.JsonException: Error creating 'OpenAI_API.ChatFunctions.FunctionCallConverter'. ---> Newtonsoft.Json.JsonException: No parameterless constructor defined for 'OpenAI_API.ChatFunctions.FunctionCallConverter'.
at Newtonsoft.Json.Serialization.JsonTypeReflector+<>c__DisplayClass22_0.<GetCreator>b__0 (System.Object[] parameters) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.JsonTypeReflector.GetJsonConverter (System.Object attributeProvider) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.DefaultContractResolver.SetPropertySettingsFromAttributes (Newtonsoft.Json.Serialization.JsonProperty property, System.Object attributeProvider, System.String name, System.Type declaringType, Newtonsoft.Json.MemberSerialization memberSerialization, System.Boolean& allowNonPublicAccess) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.DefaultContractResolver.CreateProperty (System.Reflection.MemberInfo member, Newtonsoft.Json.MemberSerialization memberSerialization) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.DefaultContractResolver.CreateProperties (System.Type type, Newtonsoft.Json.MemberSerialization memberSerialization) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.DefaultContractResolver.CreateObjectContract (System.Type objectType) [0x00000] in <00000000000000000000000000000000>:0
at System.Collections.Concurrent.ConcurrentDictionary`2[TKey,TValue].GetOrAdd (TKey key, System.Func`2[T,TResult] valueFactory) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.JsonSerializer.SerializeInternal (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.JsonConvert.SerializeObjectInternal (System.Object value, System.Type type, Newtonsoft.Json.JsonSerializer jsonSerializer) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.EndpointBase.HttpRequestRaw (System.String url, System.Net.Http.HttpMethod verb, System.Object postData, System.Boolean streaming) [0x00000] in <00000000000000000000000000000000>:0
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[TResult].Start[TStateMachine] (TStateMachine& stateMachine) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.EndpointBase.HttpRequestRaw (System.String url, System.Net.Http.HttpMethod verb, System.Object postData, System.Boolean streaming) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.EndpointBase+<HttpStreamingRequest>d__16`1[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine] (TStateMachine& stateMachine) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.EndpointBase+<HttpStreamingRequest>d__16`1[T].System.Collections.Generic.IAsyncEnumerator<T>.MoveNextAsync () [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.Chat.ChatEndpoint.StreamCompletionAsync (OpenAI_API.Chat.ChatRequest request, System.Action`2[T1,T2] resultHandler, System.Threading.CancellationToken cancellationToken) [0x00000] in <00000000000000000000000000000000>:0
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine] (TStateMachine& stateMachine) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.Chat.ChatEndpoint.StreamCompletionAsync (OpenAI_API.Chat.ChatRequest request, System.Action`2[T1,T2] resultHandler, System.Threading.CancellationToken cancellationToken) [0x00000] in <00000000000000000000000000000000>:0
at Xrai.Util.Gpt.ExecuteGptQuery (System.String prompt, System.Action`2[T1,T2] resultHandler, System.Boolean ephemeral, System.String system_message, System.Threading.CancellationToken cancellationToken) [0x00000] in <00000000000000000000000000000000>:0
at CaptionHUD.FixedUpdate () [0x00000] in <00000000000000000000000000000000>:0
--- End of inner exception stack trace ---
at Newtonsoft.Json.Serialization.JsonTypeReflector+<>c__DisplayClass22_0.<GetCreator>b__0 (System.Object[] parameters) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.JsonTypeReflector.GetJsonConverter (System.Object attributeProvider) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.DefaultContractResolver.SetPropertySettingsFromAttributes (Newtonsoft.Json.Serialization.JsonProperty property, System.Object attributeProvider, System.String name, System.Type declaringType, Newtonsoft.Json.MemberSerialization memberSerialization, System.Boolean& allowNonPublicAccess) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.DefaultContractResolver.CreateProperty (System.Reflection.MemberInfo member, Newtonsoft.Json.MemberSerialization memberSerialization) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.DefaultContractResolver.CreateProperties (System.Type type, Newtonsoft.Json.MemberSerialization memberSerialization) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.DefaultContractResolver.CreateObjectContract (System.Type objectType) [0x00000] in <00000000000000000000000000000000>:0
at System.Collections.Concurrent.ConcurrentDictionary`2[TKey,TValue].GetOrAdd (TKey key, System.Func`2[T,TResult] valueFactory) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.JsonSerializer.SerializeInternal (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00000] in <00000000000000000000000000000000>:0
at Newtonsoft.Json.JsonConvert.SerializeObjectInternal (System.Object value, System.Type type, Newtonsoft.Json.JsonSerializer jsonSerializer) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.EndpointBase.HttpRequestRaw (System.String url, System.Net.Http.HttpMethod verb, System.Object postData, System.Boolean streaming) [0x00000] in <00000000000000000000000000000000>:0
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[TResult].Start[TStateMachine] (TStateMachine& stateMachine) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.EndpointBase.HttpRequestRaw (System.String url, System.Net.Http.HttpMethod verb, System.Object postData, System.Boolean streaming) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.EndpointBase+<HttpStreamingRequest>d__16`1[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine] (TStateMachine& stateMachine) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.EndpointBase+<HttpStreamingRequest>d__16`1[T].System.Collections.Generic.IAsyncEnumerator<T>.MoveNextAsync () [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.Chat.ChatEndpoint.StreamCompletionAsync (OpenAI_API.Chat.ChatRequest request, System.Action`2[T1,T2] resultHandler, System.Threading.CancellationToken cancellationToken) [0x00000] in <00000000000000000000000000000000>:0
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine] (TStateMachine& stateMachine) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.Chat.ChatEndpoint.StreamCompletionAsync (OpenAI_API.Chat.ChatRequest request, System.Action`2[T1,T2] resultHandler, System.Threading.CancellationToken cancellationToken) [0x00000] in <00000000000000000000000000000000>:0
at Xrai.Util.Gpt.ExecuteGptQuery (System.String prompt, System.Action`2[T1,T2] resultHandler, System.Boolean ephemeral, System.String system_message, System.Threading.CancellationToken cancellationToken) [0x00000] in <00000000000000000000000000000000>:0
at CaptionHUD.FixedUpdate () [0x00000] in <00000000000000000000000000000000>:0
--- End of stack trace from previous location where exception was thrown ---

at OpenAI_API.EndpointBase+<HttpStreamingRequest>d__16`1[T].MoveNext () [0x00000] in <00000000000000000000000000000000>:0
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine] (TStateMachine& stateMachine) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.EndpointBase+<HttpStreamingRequest>d__16`1[T].System.Collections.Generic.IAsyncEnumerator<T>.MoveNextAsync () [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.Chat.ChatEndpoint.StreamCompletionAsync (OpenAI_API.Chat.ChatRequest request, System.Action`2[T1,T2] resultHandler, System.Threading.CancellationToken cancellationToken) [0x00000] in <00000000000000000000000000000000>:0
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine] (TStateMachine& stateMachine) [0x00000] in <00000000000000000000000000000000>:0
at OpenAI_API.Chat.ChatEndpoint.StreamCompletionAsync (OpenAI_API.Chat.ChatRequest request, System.Action`2[T1,T2] resultHandler, System.Threading.CancellationToken cancellationToken) [0x00000] in <00000000000000000000000000000000>:0
at Xrai.Util.Gpt.ExecuteGptQuery (System.String prompt, System.Action`2[T1,T2] resultHandler, System.Boolean ephemeral, System.String system_message, System.Threading.CancellationToken cancellationToken) [0x00000] in <00000000000000000000000000000000>:0
at CaptionHUD.FixedUpdate () [0x00000] in <00000000000000000000000000000000>:0
--- End of stack trace from previous location where exception was thrown ---
```

brondavies commented on OpenAI_API/ChatFunctions/FunctionCallConverter.cs:
```diff
@@ -0,0 +1,46 @@
+using Newtonsoft.Json;
+namespace OpenAI_API.ChatFunctions
+{
+    internal class FunctionCallConverter : JsonConverter
+    {
```
(https://github.com/OkGoDoIt/OpenAI-API-dotnet/pull/149#issuecomment-1620770517)
```suggestion
    internal class FunctionCallConverter : JsonConverter
    {
        public FunctionCallConverter() : base() { }
```

Whiteha commented:
StreamChatEnumerableAsync works in a strange way with functions. Is any working example? 

I found only the way to use it
`var conversation = api.Chat.CreateConversation(new ChatRequest
            {
                Model = Model.GPT4_0613,
                Functions = functionsList,
                Temperature = 0.10,
                Messages = messages
            });

            try
            {
                string currentFunction = "";
                var functionCalls = new Dictionary<string, string>();
                await foreach (var result in api.Chat.StreamChatEnumerableAsync(conversation.RequestParameters).ConfigureAwait(false))
                {
                    if (result.Choices[0] != null && result.Choices[0].FinishReason == "function_call")
                    {
                        foreach (var call in functionCalls)
                        {
                            // parse json and call manually
                        }
                    }
                    else if (result.Choices[0].Delta.FunctionCall?.Arguments != null)
                    {
                        // Only first entrance has function Name
                        if (result.Choices[0].Delta.FunctionCall.Name != null)
                        {
                            currentFunction = result.Choices[0].Delta.FunctionCall.Name;
                            functionCalls.Add(currentFunction, "");
                        }
                        else
                        {
                            // collecting json with arguments
                            functionCalls[currentFunction] += result.Choices[0].Delta.FunctionCall.Arguments;
                        }
                    }

                    foreach (var choice in result.Choices.Where(choice => !string.IsNullOrWhiteSpace(choice.Delta?.Content)))
                    {
                        streamGetter(choice.Delta.Content);
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }  `

OkGoDoIt commented:
Hey everyone, I'm sorry for the delay on my end.  I've been swamped with my day job and I'm behind on triaging PR's here.  I will do my best to get to this within the next few days.  Thanks for your understanding 😅

varon commented:
@OkGoDoIt - Thank you! Please check also the email I sent you!

lofcz commented on OpenAI_API/ChatFunctions/FunctionCallConverter.cs:
```diff
@@ -0,0 +1,48 @@
+    internal class FunctionCallConverter : JsonConverter
+    {
+        public FunctionCallConverter() : base() { }
+        public FunctionCallConverter() : base() { }
```
now this won't compile as it; doubled empty ctor

lofcz commented:
I advocate for friendlier exposure to the end user. For a programmer using functions together with streaming, there are three key events:

1. a message type is resolved (either `assistant` or `function`
2. if we got a `function`, it is useless to yield tokens as the entire message is needed to dispatch the requested call
3. else the current approach works (e.g. yielding tokens)

This can be implemented as three handlers for the respective events. Optionally we can support all combinations of sync / async event handlers. In the following example, I assume all event handlers are available.

```cs
/// <summary>
/// Stream LLM response. If the response is of type "function", the entire response is buffered, and then <see cref="functionCallHandler"/> is invoked.
/// Otherwise the message is intended for the end user and <see cref="messageTokenHandler"/> is called for each incoming token
/// </summary>
/// <param name="messageTokenHandler"></param>
/// <param name="functionCallHandler"></param>
/// <param name="messageTypeResolvedHandler">This is called typically after the first token arrives signaling type of the incoming message</param>
public async Task StreamResponseEnumerableFromChatbotAsyncWithFunctions(Func<string, Task> messageTokenHandler, Func<string, Task> functionCallHandler, Func<ChatMessageRole, Task> messageTypeResolvedHandler)
{
	ChatRequest req = new ChatRequest(RequestParameters)
	{
		Messages = _Messages.ToList()
	};

	StringBuilder responseStringBuilder = new StringBuilder();
	ChatMessageRole responseRole = null;

	await foreach (ChatResult res in _endpoint.StreamChatEnumerableAsync(req))
	{
		if (res.Choices.FirstOrDefault()?.Delta is { } delta)
		{
			string deltaContent = delta.Content;
			bool empty = string.IsNullOrEmpty(deltaContent);

			if (responseRole != null && responseRole.Equals(ChatMessageRole.Function))
			{
				if (!empty)
				{
					responseStringBuilder.Append(deltaContent);
				}
				
				continue;
			}
			
			if (responseRole == null && delta.Role != null)
			{
				responseRole = delta.Role;

				if (messageTypeResolvedHandler != null)
				{
					await messageTypeResolvedHandler(responseRole);	
				}

				if (functionCallHandler != null && responseRole == "function")
				{
					if (!empty)
					{
						responseStringBuilder.Append(deltaContent);
					}
					
					continue;
				}
			}
                  
			if (!empty)
			{
				responseStringBuilder.Append(deltaContent);

				if (messageTokenHandler != null)
				{
					await messageTokenHandler.Invoke(deltaContent);	
				}
			}
		}
		
		MostRecentApiResult = res;
	}

	if (responseRole != null && responseRole.Equals(ChatMessageRole.Function))
	{
		if (functionCallHandler != null)
		{
			await functionCallHandler.Invoke(responseStringBuilder.ToString());
		}
	}

	if (responseRole != null)
	{
		AppendMessage(responseRole, responseStringBuilder.ToString());
	}
}
```

@hansjm10 @OkGoDoIt thoughts?

lofcz commented:
Efficient implementation of function calling in conjunction with streaming: https://github.com/lofcz/OpenAI-API-dotnet/commit/52b78842d8d223441732f892b29d0d1578b61c0f#diff-8bd7aaf5181bcc44301c5b518a91bee8f7e4591461c1c19dc8a4a55e74354e07R383

ffMathy commented:
@OkGoDoIt can you provide us with an additional update on this?

iwaitu commented:
@hansjm10 I've fixed the issue where function_call couldn't be used properly during re-streaming. Now, function_call can be used normally in all scenarios. [detail](https://github.com/hansjm10/OpenAI-API-dotnet/compare/master...iwaitu:OpenAI-API-dotnet:master).Also, I've added example code for using stream transmission. It's located in test/ChatEndpointTest.cs, within the method named [SummarizeFunctionStreamResult](https://github.com/hansjm10/OpenAI-API-dotnet/compare/master...iwaitu:OpenAI-API-dotnet:master#diff-7f034c76bfb15fdf6a65b325c5af1256e94bb98c4202b8dc9d7a16f0254b6826).

hansjm10 commented:
> @hansjm10 I've fixed the issue where function_call couldn't be used properly during re-streaming. Now, function_call can be used normally in all scenarios. [detail](https://github.com/hansjm10/OpenAI-API-dotnet/compare/master...iwaitu:OpenAI-API-dotnet:master).Also, I've added example code for using stream transmission. It's located in test/ChatEndpointTest.cs, within the method named [SummarizeFunctionStreamResult](https://github.com/hansjm10/OpenAI-API-dotnet/compare/master...iwaitu:OpenAI-API-dotnet:master#diff-7f034c76bfb15fdf6a65b325c5af1256e94bb98c4202b8dc9d7a16f0254b6826).

Looks good and I'm ready to merge it but just revert the changes to the proj files.

ErikDombi commented:
There's a lot of weird intending issues in a lot of the files as well

brondavies commented:
> There's a lot of weird intending issues in a lot of the files as well

I think @ErikDombi means "indenting issues" as in whitespace, and formatting is all over the place 😄

MathiasGronseth commented:
Is there any status on this? Eager to try out this functionality

DaliaDawod commented:
I have this and token is printed empty, can someone help me?

      ```
 var functionParameters =  new{
            type = "object",
            properties = new
            {
                scary = new
                {
                    type = "string",
                    description = "A scary version of the response to a user's query"
                },
                joyful = new
                {
                    type = "string",
                    description = "A joyful version of the response to a user's query"
                }
            },
            required = new[]
            {
                "scary",
                "joyful"
            }
            
        };

        var functions = new List<Function> { new Function(
            "responses",
            "answer in an angry way",
            functionParameters
        )};

        var functionCall = new FunctionCall
        {
            Name = "responses" ,
            // Arguments = "scary"
        };


        // Send the entire chat to OpenAI to get the next message
        await foreach (var token in api.Chat.StreamChatEnumerableAsync(new ChatRequest()
        {
                Model = Model.ChatGPTTurbo,
                Temperature = 0.5,
                Messages = messages,
                Functions = functions,  
                FunctionCall = functionCall
        }))
        {
        Console.WriteLine("token ", token);
.......
```

EvoTechMike commented:
Looks like PR got killed.  Can we get it back?

EvoTechMike commented:
did it get merged?

OkGoDoIt commented:
@EvoTechMike, no, I haven't merged any PR's in a while. I implemented some of this myself but I have not yet implemented the function calling.

lionbarz commented:
Function calling is the only reason I can't use this beautiful package.

endeffects commented:
@OkGoDoIt If you don't have time to implement or approve features by yourself why you don't let your contributors do the job? Still if a community implementation doesn't  reach your quality standards and will be replaced by your own ones later on it would be a much better progress. Right now we have a feature lack which grows from day to day. I understand that you have your own personal goals but on a project that became so famous there is a requirement to be up to date and feature complete. So please rethink about your future project management.

lofcz commented:
Interim one of the forks supporting Functions, Assistants and other new features can be used.

https://github.com/RageAgainstThePixel/OpenAI-DotNet - supports Unity, .NET Standard & .NET Core, frequently updated, great test coverage
https://github.com/lofcz/OpenAiNg - .NET Core only, high performance, customizability, supports locally hosted models, responses include raw http request.
(Ordered by ⭐️ at the time of writing)

endeffects commented:
Thanks

EvoTechMike commented:
@lofcz Out of curiosity, how did you find these forks?  When I go to https://github.com/OkGoDoIt/OpenAI-API-dotnet/forks there isn't any forks with more than 1 star.

lofcz commented:
Both RageAgainstThePixel and me hardforked some time ago. I've maintained a soft fork for about half a year but then the changes diverged too much from upstream, hence hard fork.

cnSchwarzer commented:
I suggest archive this repository if the author no longer maintain this feature. It is funny that such feature is not merged after 8 months.
